cot/src/project.rs

```diff
@@ -87,605 +155,1439 @@ pub trait CotApp: Send + Sync {
+    ///     fn register_apps(
+    ///         &self,
+    ///         apps: &mut cot::AppBuilder,
+    ///         _context: &cot::ProjectContext<WithConfig>,
+    ///     ) {
+    ///         apps.register_with_views(HelloApp, "/hello");
+    ///     }
+    /// }
+    /// ```
+    pub fn register_with_views<T: App + 'static>(&mut self, module: T, url_prefix: &str) {
+        self.urls
+            .push(Route::with_router(url_prefix, module.router()));
+        self.register(module);
+    }
+}
+
+/// A trait for defining custom error page handlers.
+///
+/// This is useful with [`Project::server_error_handler`] and
+/// [`Project::not_found_handler`].
+///
+/// # Examples
+///
+/// ```
+/// use cot::project::ErrorPageHandler;
+/// use cot::response::{Response, ResponseExt};
+/// use cot::{Body, Project, StatusCode};
+///
+/// struct MyProject;
+/// impl Project for MyProject {
+///     fn not_found_handler(&self) -> Box<dyn ErrorPageHandler> {
+///         Box::new(MyHandler)
+///     }
+/// }
+///
+/// struct MyHandler;
+/// impl ErrorPageHandler for MyHandler {
+///     fn handle(&self) -> cot::Result<Response> {
+///         Ok(Response::new_html(
+///             StatusCode::NOT_FOUND,
+///             Body::fixed("Not Found"),
+///         ))
+///     }
+/// }
+/// ```
+pub trait ErrorPageHandler: Send + Sync {
+    /// Returns the error response.
+    ///
+    /// # Errors
+    ///
+    /// This method may return an error if the handler fails to build a
+    /// response. In this case, the error will be logged and a generic
+    /// error page will be returned to the user.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use cot::project::ErrorPageHandler;
+    /// use cot::response::{Response, ResponseExt};
+    /// use cot::{Body, Project, StatusCode};
+    ///
+    /// struct MyProject;
+    /// impl Project for MyProject {
+    ///     fn not_found_handler(&self) -> Box<dyn ErrorPageHandler> {
+    ///         Box::new(MyHandler)
+    ///     }
+    /// }
+    ///
+    /// struct MyHandler;
+    /// impl ErrorPageHandler for MyHandler {
+    ///     fn handle(&self) -> cot::Result<Response> {
+    ///         Ok(Response::new_html(
+    ///             StatusCode::NOT_FOUND,
+    ///             Body::fixed("Not Found"),
+    ///         ))
+    ///     }
+    /// }
+    /// ```
+    fn handle(&self) -> crate::Result<Response>;
+}
+
+struct DefaultNotFoundHandler;
+impl ErrorPageHandler for DefaultNotFoundHandler {
+    fn handle(&self) -> crate::Result<Response> {
+        Ok(Response::new_html(
+            StatusCode::NOT_FOUND,
+            Body::fixed(include_str!("../templates/404.html")),
+        ))
+    }
+}
+
+struct DefaultServerErrorHandler;
+impl ErrorPageHandler for DefaultServerErrorHandler {
+    fn handle(&self) -> crate::Result<Response> {
+        Ok(Response::new_html(
+            StatusCode::INTERNAL_SERVER_ERROR,
+            Body::fixed(include_str!("../templates/500.html")),
```

Comment:
Can you reuse `DEFAULT_SERVER_ERROR_PAGE` and `DEFAULT_NOT_FOUND_PAGE` instead of embedding them again?

Reply:
This would require exposing both these constants and the `error_page` module as `pub(crate)` which I don't want to do if not needed (to promote separation of responsibilities in the project).

I also think that we could potentially want to separate these two default templates and add something like "your custom error page failed to render, that's why you see this simplified error instead" to `DEFAULT_SERVER_ERROR_PAGE` and `DEFAULT_NOT_FOUND_PAGE` in the future. Let's keep them (sort of) separated for now. Rust compiler should be smart enough only to include one copy in the binary.